test/integration/BCoWHelper.t.sol

```diff
@@ -0,0 +1,229 @@
+import {BCoWHelper} from 'contracts/BCoWHelper.sol';
+
+contract ConstantProductHelperForkedTest is Test {
+  using GPv2Order for GPv2Order.Data;
+
+  BCoWHelper private helper;
+
+  // All hardcoded addresses are mainnet addresses
+  address public lp = makeAddr('lp');
+
+  ISettlement private settlement = ISettlement(0x9008D19f58AAbD9eD0D60971565AA8510560ab41);
+  address private vaultRelayer;
+
+  address private solver = 0x423cEc87f19F0778f549846e0801ee267a917935;
+
+  BCoWFactory private ammFactory;
+  IBPool private weightedPool;
+  IBPool private basicPool;
+
+  IERC20 private constant DAI = IERC20(0x6B175474E89094C44Da98b954EedeAC495271d0F);
+  IERC20 private constant USDC = IERC20(0xA0b86991c6218b36c1d19D4a2e9Eb0cE3606eB48);
+  IERC20 private constant WETH = IERC20(0xC02aaA39b223FE8D0A0e5C4F27eAD9083C756Cc2);
+
+  uint256 constant VALID_AMOUNT = 1e6;
+  uint256 constant TEN_PERCENT = 0.1 ether;
+
+  function setUp() public {
+    vm.createSelectFork('mainnet', 20_012_063);
+
+    vaultRelayer = address(settlement.vaultRelayer());
+
+    ammFactory = new BCoWFactory(address(settlement), bytes32('appData'));
+    helper = new BCoWHelper(address(ammFactory));
+
+    deal(address(DAI), lp, 2 * VALID_AMOUNT);
+    deal(address(WETH), lp, 2 * VALID_AMOUNT);
+
+    vm.startPrank(lp);
+    weightedPool = ammFactory.newBPool();
+    basicPool = ammFactory.newBPool();
+
+    DAI.approve(address(weightedPool), type(uint256).max);
+    WETH.approve(address(weightedPool), type(uint256).max);
+    weightedPool.bind(address(DAI), VALID_AMOUNT, 8e18); // 80% weight
+    weightedPool.bind(address(WETH), VALID_AMOUNT, 2e18); // 20% weight
+
+    DAI.approve(address(basicPool), type(uint256).max);
+    WETH.approve(address(basicPool), type(uint256).max);
+    basicPool.bind(address(DAI), VALID_AMOUNT, 4.2e18); // no weight
+    basicPool.bind(address(WETH), VALID_AMOUNT, 4.2e18); // no weight
+
+    // finalize
+    weightedPool.finalize();
+    basicPool.finalize();
+
+    vm.stopPrank();
+  }
+
+  // NOTE: 1 ETH = 1000 DAI
+  uint256 constant INITIAL_SPOT_PRICE = 0.001e18;
+
+  function testBasicOrder() public {
+    IBCoWPool pool = IBCoWPool(address(basicPool));
+
+    uint256 ammWethInitialBalance = 1 ether;
+    uint256 ammDaiInitialBalance = 1000 ether;
+
+    deal(address(WETH), address(pool), ammWethInitialBalance);
+    deal(address(DAI), address(pool), ammDaiInitialBalance);
+
+    uint256 spotPrice = pool.getSpotPriceSansFee(address(WETH), address(DAI));
+    assertEq(spotPrice, INITIAL_SPOT_PRICE);
+
+    _executeHelperOrder(pool, ammWethInitialBalance, ammDaiInitialBalance);
+
+    uint256 postSpotPrice = pool.getSpotPriceSansFee(address(WETH), address(DAI));
+    assertEq(postSpotPrice, 1_052_631_578_947_368);
+  }
+
+  // NOTE: this test checks that the order generated by the helper is valid
+  function testValidOrder(uint256 balanceToken0, uint256 balanceToken1, uint256 priceSkewness) public {
```

Comment:
I believe it would be best to have a hard-coded valid scenario (optionally with data as close to current blockchain state possible) in integration tests and perform fuzzing in unit tests, since they are executed with a higher `FOUNDRY_FUZZ_RUNS` parameter, and the contracts implementing the fuzzed logic are all offline-available libraries

Reply:
hmmm, i'm not entirely happy with fuzzed forked tests, but we do have little expectations, and we're trying to test a behaviour between multiple contracts: that what the helper returns is valid in the pool, so in a way yes, we could add the pool to the helper unit test contract and do it there, in another way i feel it could live within the integration scope, what do you think?

Reply:
we're actually not using any oracle here either, it could be done all within a unit test scenario, but indeed requires more than 1 contract and it doesn't fit any tree